OTAnalytics/application/progress.py

```diff
@@ -0,0 +1,78 @@
+from typing import Any, Callable, Iterator, Optional, Sequence
+
+from OTAnalytics.domain.progress import Counter, Progressbar
+
+
+class SimpleCounter(Counter):
+    def __init__(self) -> None:
+        self._value = 0
+
+    def increment(self, value: int = 1) -> None:
+        self._value += value
+
+    def reset(self) -> None:
+        self._value = 0
+
+    def get_value(self) -> int:
+        return self._value
+
+
+class AutoIncrementingProgressbar(Progressbar):
+    def __init__(
+        self,
+        sequence: Sequence,
+        counter: Counter,
+        notify: Optional[Callable[[], None]] = None,
+    ) -> None:
+        self._sequence = sequence
+        self._counter = counter
+        self._notify = notify
+        self._iterator = iter(self._sequence)
+        self._counter.reset()
+
+    def __iter__(self) -> Iterator:
+        self._counter.reset()
+        self._iterator = iter(self._sequence)
+        while True:
+            try:
+                yield next(self._iterator)
+                self._counter.increment(1)
+                if self._notify:
+                    self._notify()
+            except StopIteration:
+                return
+
+
+class ManualIncrementingProgressbar(Progressbar):
+    def __init__(
+        self,
+        sequence: Sequence,
+        counter: Counter,
+        notify: Optional[Callable[[], None]] = None,
+    ) -> None:
+        self._sequence = sequence
+        self._counter = counter
+        self._notify = notify
+        self._iterator = iter(self._sequence)
+        self._counter.reset()
+
+    def __iter__(self) -> Iterator:
+        self._counter.reset()
+        self._iterator = iter(self._sequence)
+        return self
+
+    def __next__(self) -> Any:
+        next_element = next(self._iterator)
+        if self._notify:
+            self._notify()
+        return next_element
+
```

Comment:
Why do we need a `__next__` method here but not in the automatic version?

Reply:
The automatic version is basically a `Generator` in disguise. And every `Generator`  is an `Iterator` Whereas the manual version implements the `Iterator` interface.
I could have implemented the manual version as a generator as well. The reason why I didn't was due to the fact that `Generator`s are not pickable which is needed for `multiprocessing`.